File: Common/Network/Messages/LaunchConfiguration/LaunchConfigurationResponse.cs
namespace Common.Network.Messages.LaunchConfiguration;

public class LaunchConfigurationResponse: Message
{
    public Common.Models.LaunchConfiguration LaunchConfiguration;

    public LaunchConfigurationResponse()
    {
    }

    public LaunchConfigurationResponse(Common.Models.LaunchConfiguration launchConfiguration)
    {
        LaunchConfiguration = launchConfiguration;
    }

    public override MessageHeader GetHeader()
    {
        return new MessageHeader(MessageRegistry.GetKeyForMessageType(typeof(LaunchConfigurationResponse)), 
            LaunchConfiguration.ForgeVersion.Length + 
            LaunchConfiguration.ModsChecksum.Length + 
            LaunchConfiguration.VanillaVersion.Length + sizeof(int) * 4);    
    }

    public override void ApplyData(Stream stream)
    {
        LaunchConfiguration = new Common.Models.LaunchConfiguration(
            StringReadStream(stream),
            StringReadStream(stream),
            StringReadStream(stream),
            IntReadStream(stream));
    }

    protected override Stream GetData()
    {
        MemoryStream buffer = new MemoryStream(GetHeader().DataLength);
        WriteToStream(buffer, LaunchConfiguration.VanillaVersion);
        WriteToStream(buffer, LaunchConfiguration.ForgeVersion);
        WriteToStream(buffer, LaunchConfiguration.ModsChecksum);
        WriteToStream(buffer, LaunchConfiguration.RequiredGriwnas);
        
        buffer.Flush();
        buffer.Position = 0;
        return buffer;
    }
}

File: IntegrationMod/src/main/java/com/example/examplemod/bfml/LaunchConfigurationResponse.java
package com.example.examplemod.bfml;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public final class LaunchConfigurationResponse extends Message {

    public String ModsChecksum;
    public String VanillaVersion;
    public String ForgeVersion;
    public int RequiredGriwnas;

    @Override
    public MessageHeader getMessageHeader() throws Exception {
        return new MessageHeader(MessageRegistry.getKeyForMessageType(LaunchConfigurationResponse.class),
                4 * (Integer.SIZE / 8) + VanillaVersion.length() + ForgeVersion.length() + ModsChecksum.length());
    }

    @Override
    public void applyData(InputStream stream) throws IOException {
        VanillaVersion = stringReadStream(stream);
        ForgeVersion= stringReadStream(stream);
        ModsChecksum = stringReadStream(stream);
        RequiredGriwnas = intReadStream(stream);
    }

    @Override
    public byte[] getData() throws Exception {
        MessageHeader header = getMessageHeader();
        ByteBuffer byteBuffer = ByteBuffer.allocate(header.DataLength);
        byteBuffer.order(ByteOrder.LITTLE_ENDIAN);
        byteBuffer.putInt(VanillaVersion.length());
        byteBuffer.put(VanillaVersion.getBytes());
        byteBuffer.putInt(ForgeVersion.length());
        byteBuffer.put(ForgeVersion.getBytes());
        byteBuffer.putInt(ModsChecksum.length());
        byteBuffer.put(ModsChecksum.getBytes());
        byteBuffer.putInt(RequiredGriwnas);

        return byteBuffer.array();
    }
}
